File: lib/wreq_ruby/tls.rb
# frozen_string_literal: true

unless defined?(Wreq)
  module Wreq
    # Peer certificate data captured for one HTTPS response.
    #
    # Instances are returned by {Wreq::Response#tls_info}. Certificate bytes
    # remain available after the response body is read or closed, even if the
    # connection is later reused.
    #
    # The returned certificate Strings are Ruby-owned copies. Changing one does
    # not alter the stored TLS data or values returned by later calls. The chain
    # Array is frozen, but its String elements remain mutable.
    #
    # Certificates use the DER encoding described by the X.509 profile in
    # RFC 5280.
    #
    # @example Parse the peer certificate with OpenSSL
    #   require "openssl"
    #
    #   client = Wreq::Client.new(tls_info: true)
    #   response = client.get("https://example.com")
    #   der = response.tls_info&.peer_certificate
    #
    #   if der
    #     certificate = OpenSSL::X509::Certificate.new(der)
    #     puts certificate.subject
    #   end
    # @see https://www.rfc-editor.org/rfc/rfc5280#section-4.1 X.509 certificate format
    class TlsInfo
      # Return the peer's leaf certificate.
      #
      # @return [String, nil] a new DER-encoded String with
      #   +Encoding::BINARY+, or +nil+ when the transport did not provide one
      def peer_certificate
      end

      # Return the peer certificate chain.
      #
      # The Array is frozen. Each element is a new DER-encoded binary String.
      # The chain includes the leaf certificate when the transport supplies it.
      #
      # @return [Array<String>, nil] a frozen Array of certificate copies, or
      #   +nil+ when the transport did not provide a chain
      def peer_certificate_chain
      end
    end
  end
end

# ======================== Ruby API Extensions ========================

module Wreq
  class TlsInfo
    # Return a compact summary for debugging.
    #
    # The summary reports the leaf certificate size and the number of
    # certificates in the chain without printing the DER data.
    #
    # @return [String] TLS certificate metadata
    # @example
    #   tls_info.inspect
    #   # => "#<Wreq::TlsInfo peer_certificate=781B peer_certificate_chain=1>"
    def inspect
      certificate = peer_certificate
      chain = peer_certificate_chain
      certificate_size = certificate ? "#{certificate.bytesize}B" : "nil"
      chain_size = chain ? chain.length : "nil"

      "#<#{self.class} peer_certificate=#{certificate_size} peer_certificate_chain=#{chain_size}>"
    end
  end
end

File: src/tls.rs
//! Ruby wrappers for TLS metadata attached to a response.
//!
//! Certificates use the DER encoding described by the X.509 profile in
//! [RFC 5280 section 4.1](https://www.rfc-editor.org/rfc/rfc5280#section-4.1).

use magnus::{Error, Module, RArray, RModule, RString, Ruby, value::ReprValue};

/// Read-only Ruby wrapper around [`wreq::tls::TlsInfo`].
///
/// The native value keeps certificate bytes alive independently of the response
/// body. Its `Bytes` buffers are cheap to clone, while accessors copy the data
/// into Ruby-owned Strings so callers cannot mutate the stored metadata.
#[derive(Clone)]
#[magnus::wrap(class = "Wreq::TlsInfo", free_immediately, size)]
pub(crate) struct TlsInfo(pub(crate) wreq::tls::TlsInfo);

impl TlsInfo {
    /// Copy the DER-encoded leaf certificate into a binary Ruby String.
    fn peer_certificate(ruby: &Ruby, rb_self: &Self) -> Option<RString> {
        rb_self
            .0
            .peer_certificate()
            .map(|der| ruby.str_from_slice(der))
    }

    /// Copy the certificate chain into a frozen Array of binary Ruby Strings.
    ///
    /// Only the Array is frozen. Its Strings are independent copies and remain
    /// mutable in Ruby.
    fn peer_certificate_chain(ruby: &Ruby, rb_self: &Self) -> Option<RArray> {
        rb_self.0.peer_certificate_chain().map(|chain| {
            let certificates = ruby.ary_from_iter(chain.map(|cert| ruby.str_from_slice(cert)));
            certificates.freeze();
            certificates
        })
    }
}

/// Define the `Wreq::TlsInfo` Ruby class and its readers.
pub(crate) fn include(ruby: &Ruby, gem_module: &RModule) -> Result<(), Error> {
    let tls_info_class = gem_module.define_class("TlsInfo", ruby.class_object())?;
    tls_info_class.define_method(
        "peer_certificate",
        magnus::method!(TlsInfo::peer_certificate, 0),
    )?;
    tls_info_class.define_method(
        "peer_certificate_chain",
        magnus::method!(TlsInfo::peer_certificate_chain, 0),
    )?;
    Ok(())
}
